### ISSUE TITLE
SSR and cache-and-network fetch policy

### ISSUE THREAD

**Issue by mxstbr** [apollographql/apollo-client#2119]
**Intended outcome:**

I server-side render my React application. I used `renderToStringWithData` on the backend to get all the data, then serialized the state with Yahoos `serialize-javascript` into the generated HTML and then added it to my custom Redux store as the initial state via `window.__SERVER_STATE__`.

I expected the client bundle to load in and not fetch anything.

**Actual outcome:**

The client bundle loads in and React rehydrates, then Apollo fetches `/api` even though it already has all the data from the initial state in the HTML.

This seems backwards, since even with `cache-and-network` after the SSR'd first result I wouldn't expect apollo client to go to the network again on first render.

**How to reproduce the issue:**

- Have an SSR'd React app
- Use a fetch policy of `cache-and-network`
- Check your network requests

**Version**

- apollo-client@1.9.1

**Comment by mhuggins** [comment]
I just started using Apollo and am seeing this as well.  Not sure if there's an existing solution or if this is a bug.

**Comment by mhuggins** [comment]
As a workaround, I'm now using the `ssr: false` option to prevent loading data on the server, instead just sending the unloaded/loading state back to the client.  http://dev.apollodata.com/react/server-side-rendering.html#skip-for-ssr

**Comment by mhuggins** [comment]
This is still an issue.

**Comment by mxstbr** [comment]
Definitely still an issue.

**Comment by leoasis** [comment]
Isn't this expected behavior with `cache-and-network`? According to the docs here http://dev.apollodata.com/react/api-queries.html#graphql-config-options-fetchPolicy, that policy will always execute a network request even if the data is present in the cache (in which case that data is also returned). Maybe the `cache-first` policy would have the behavior you'd expect?

**Comment by mxstbr** [comment]
No. I want my query to run again when a user visits a page again to get the freshest data, but I also want it to show the stale data for a faster render and nicer UX.

It's just now rendered with the fresh data on the server, why would I want it to stress my server by unnecessarily fetching the exact same queries that were just loaded literally milliseconds ago?

**Comment by leoasis** [comment]
Yeah I understand that, just wanted to say that it was kind of the expected behavior IMO, because the component is returning the data from the cache and _also_ firing a network request, as it is documented. The problem though is that in this case, the cache was populated by that same component on the server side, but that state is not transferred to the client (since it's not part of the store), only the data in the store.

Maybe there could be a way to set the fetchPolicy to `cache-first` and as soon as the component is filled with the data, it could be changed back to `cache-and-network`?

To make this work "transparently" the client would have to also save its internal state for the queries being watched, and have some way to rehydrate those based on the components that mount. This looks like it would be very hard to implement.

Another solution would be to ignore the fetch policy when hydrating from the server, and just populate the data from the cache. Something similar to `ssrForceFetchDelay`.

**Comment by jbaxleyiii** [comment]
@mxstbr so you are saying the initial data from the SSR isn't present and it resets and makes a network request?

Looking at the intended outcome from the issue, you don't want it to make the request again but that is what that policy is designed to do? I'm not sure I understand what is wrong?

**Comment by mxstbr** [comment]
No, the data is there, that's the whole point.

1. User requests `/something`
2. We render that page on the server with `renderToStringWithData` and send down the full HTML with all the data
3. The user gets the full HTML and downloads and executes the JS bundle
4. Now React rehydrates and Apollo rehydrates too
5. At this point, the user has the full app and the JS bundle has rehydrated
6. Apollo goes and fetches the data we already have from the SSR initial state again

That sixth step doesn't make any sense: the server just loaded the data and sent it down, why would Apollo go back to the server to get the exact same data again?!

**Comment by jbaxleyiii** [comment]
@mxstbr that totally shouldn't be happening! Does that only happen with certain fetch policies? Can you share your initialization code and SSR code where you inline the state? I've got a few SSR apps and they don't double request so I'm looking to narrow down the issue!

**Comment by mxstbr** [comment]
It only happens with `cache-and-network`, so it seems to be semi-intended behaviour but to me that's wrong behaviour since it doesn't make any sense to give my server double the load. 😅

You can replicate this in your own browser by opening `https://spectrum.chat/styled-components` in a private browser window (make sure to clear your browser cache each time since we cache the app shell locally with service workers) and looking at the network tab. You'll get the SSR'd HTML, but then after the JS bundle loads you'll see a request to `/api` that loads the entire data again:

<img width="1552" alt="screen shot 2017-10-10 at 8 45 18 pm" src="https://user-images.githubusercontent.com/7525670/31404310-f7ee92a8-adfb-11e7-8687-47783f4dfeb0.png">

> Note: Sorry for the german browser interface, Firefox Nightly for some reason doesn't use the OS language but is based on location or something 🤷‍♂️ 

Client initialisation:

```JS
const store = initStore(window.__SERVER_STATE__ || initialState, {
  middleware: [client.middleware()],
  reducers: {
    apollo: client.reducer(),
  },
});

function render() {
  return ReactDOM.render(
    <ApolloProvider store={store} client={client}>
      <Router history={history}>
        <Routes
          maintenanceMode={process.env.REACT_APP_MAINTENANCE_MODE === 'enabled'}
        />
      </Router>
    </ApolloProvider>,
    document.querySelector('#root')
  );
}
```

Client creation:

```JS
const networkInterface = createBatchingNetworkInterface({
  uri: API_URI,
  batchInterval: 10,
  opts: {
    credentials: 'include',
  },
});

const networkInterfaceWithSubscriptions = addGraphQLSubscriptions(
  networkInterface,
  wsClient
);

const fragmentMatcher = new IntrospectionFragmentMatcher({
  introspectionQueryResultData,
});

export const client = new ApolloClient({
  networkInterface: networkInterfaceWithSubscriptions,
  fragmentMatcher,
  initialState: window.__SERVER_STATE__ && {
    apollo: window.__SERVER_STATE__.apollo,
  },
  ssrForceFetchDelay: 100,
  ...getSharedApolloClientOptions(),
});
```

Server (express.js) route:

```JS
const renderer = (req, res) => {
  // apollo
  const client = new ApolloClient({
    ssrMode: true,
    networkInterface: createLocalInterface(graphql, schema, {
      context: {
        loaders: createLoaders(),
        user: req.user,
      },
    }),
    ...getSharedApolloClientOptions(),
  });
  // redux
  const initialReduxState = {
    users: {
      currentUser: req.user,
    },
  };

  const store = initStore(initialReduxState, {
    // Inject the server-side client's middleware and reducer
    middleware: [client.middleware()],
    reducers: {
      apollo: client.reducer(),
    },
  });
  const context = {};

  // Frontend
  const frontend = (
    <ApolloProvider store={store} client={client}>
      <StaticRouter location={req.url} context={context}>
        <Routes maintenanceMode={IN_MAINTENANCE_MODE} />
      </StaticRouter>
    </ApolloProvider>
  );
  // styled-components
  const sheet = new ServerStyleSheet();
  renderToStringWithData(sheet.collectStyles(frontend))
    .then(content => {
      if (context.url) {
        res.redirect(301, context.url);
        return;
      }
      const state = store.getState();
      // react-helmet for meta tags
      const helmet = Helmet.renderStatic();
      res.status(200);
      // Compile the HTML and send it down
      res.send(
        constructHTML({
          content,
          state,
          styleTags: sheet.getStyleTags(),
          metaTags:
            helmet.title.toString() +
            helmet.meta.toString() +
            helmet.link.toString(),
        })
      );
      res.end();
    })
    .catch(err => {
      console.log(err);
      res.status(500);
      res.end();
      throw err;
    });
};
```

**Comment by mxstbr** [comment]
Argg, not stale. Thank god I can label it myself now...

**Comment by mxstbr** [comment]
Wait that still doesn't work, wat

**Comment by mhuggins** [comment]
Still not stale.

**Comment by particle4dev** [comment]
@mxstbr What version of apollo you use ? I have same issue with apollo-client@2.1.0 then I realized  that I did not migrated code from 1.0 correctly. 
Change 
```
     return new ApolloClient({
      cache,
      initialState,
```
to
```
     return new ApolloClient({
      cache: cache.restore(initialState),
```
solved my problem. Hope this help :)

**Comment by mxstbr** [comment]
I'm on 1.0

**Comment by benjie** [comment]
This issue persists with Apollo 2.0 if you use `cache-and-network` fetch policy, e.g.:
```js
const client = new ApolloClient({
  link,
  cache: new InMemoryCache().restore(window.__APOLLO_STATE__),
  defaultOptions: {
    query: {
      fetchPolicy: "cache-and-network",
    },
  },
  ssrForceFetchDelay: 1000,
});
```

It would be nice to be able to use "cache-first" as the network policy for the first second after SSR render, then switch to "cache-and-network" after - an option like `ssrCacheFirstFetchPolicyDuration: 1000` might help?

**Comment by benjie** [comment]
It seems like the issue could be solved around here:

https://github.com/apollographql/apollo-client/blob/337314111bc75e17834d0dfb6af461b921276dd2/packages/apollo-client/src/ApolloClient.ts#L209-L211

Maybe something like this would work:

```js
    if (this.disableNetworkFetches && (options.fetchPolicy === 'network-only' || options.fetchPolicy === 'cache-and-network')) {
      options = { ...options, fetchPolicy: 'cache-first' } as WatchQueryOptions;
    }
```

Would a PR performing this change be considered?

**Comment by Alivejke** [comment]
Seems like an easy fix for a long open issue

**Comment by benjie** [comment]
I'd be more than happy to fix this in a PR if I can get a collaborator to just give me the okay.

**Comment by jbaxleyiii** [comment]
@benjie go for it!

**Comment by alexandcote** [comment]
@benjie any update on this ?

**Comment by benjie** [comment]
Sorry I’ve not had time :( A potential fix is detailed above though, feel free to give it a go yourself.

**Comment by ciscorn** [comment]
#3372 should be reviewed and merged.

**Comment by hwillson** [comment]
@ciscorn It will be reviewed shortly (it's on my radar for hopefully later today). Thanks!

**Comment by vincentbello** [comment]
Was this ever fixed? I'm still getting the exact same issue @mxstbr was reporting, where cache-and-network queries refetch once the client bundle loads in.
Is there a way to ignore the fetch policy when hydrating from the server and populate data from the cache?

**Comment by 6axter82** [comment]
Is it still open or merged?

**Comment by sgup** [comment]
Having this issue too

**Comment by geraldstrobl** [comment]
Has anyone a workaround for this issue?

**Comment by tomraithel** [comment]
The solution to this problem is setting an appropriate `ssrForceFetchDelay`:

> If you are using `fetchPolicy: network-only` or `fetchPolicy: cache-and-network` on some of the initial queries, you can pass the `ssrForceFetchDelay` option to skip force fetching during initialization, so that even those queries run using the cache:

As stated on the [Apollo SSR docu](https://www.apollographql.com/docs/react/performance/server-side-rendering/#store-rehydration).

**Comment by Moniv9** [comment]
Is this resolved? As same is still reproducible.

**Comment by darrylmabini** [comment]
Boom! using `ssrForceFetchDelay` do the trick.
